org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/internal/launch/BuildExecutionParticipants.java

```diff
@@ -5,49 +5,65 @@
  * which accompanies this distribution, and is available at
  * http://www.eclipse.org/legal/epl-v10.html
  *
- * Contributors:
- *     Etienne Studer & Donát Csikós (Gradle Inc.) - initial API and implementation and initial documentation
- *     Simon Scholz <simon.scholz@vogella.com> - Bug 478054
  */
 
 package org.eclipse.buildship.core.internal.launch;
 
-import org.eclipse.buildship.core.internal.CorePlugin;
-import org.eclipse.core.runtime.IConfigurationElement;
-import org.eclipse.core.runtime.Platform;
-import org.osgi.framework.Bundle;
-import org.osgi.framework.BundleException;
+import java.util.Collections;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.ExecutionException;
+import java.util.stream.Collectors;
 
-/**
- * Helper class to activate plugins which contribute to the {@code executionparticipants} extension point.
- */
-final class BuildExecutionParticipants {
+import org.eclipse.buildship.core.BuildExecutionParticipant;
+import org.eclipse.buildship.core.internal.extension.BuildExecutionParticipantContribution;
+import org.eclipse.buildship.core.internal.extension.InternalBuildExecutionParticipant;
 
-    // the fully-qualified name of the extension point
-    private static final String EXECUTION_PARTICIPANTS_EXTENSION_ID = CorePlugin.PLUGIN_ID + ".executionparticipants";
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
+import com.google.common.collect.Lists;
 
-    // the attribute specifying the plugin id to start
-    private static final String EXTENSION_ATTRIBUTE_PLUGIN_ID = "id";
+public class BuildExecutionParticipants {
 
-    private BuildExecutionParticipants() {
-    }
+    private static BuildExecutionParticipants instance;
+
+    private List<InternalBuildExecutionParticipant> contributions = Lists.newArrayList();
 
-    public static void activateParticipantPlugins() {
```

Comment:
We would still need this code right?  for those that just extend the 'plugin'.  We need to put that back, and just add the 'buildexecutionparticipants'